push: improve 'no remote specified' hint

fabiosantoscode opened the pull request:
This tries to clear up the error message of NoRemoteError in case the user doesn't have any remotes, and provides instructions on how to create a remote, or a default remote. It also clears the extra newline at the end.

Fixes #3121

----

* [x] ❗ Have you followed the guidelines in the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) list?

* [x] 📖 Check this box if this PR **does not** require [documentation](https://dvc.org/doc) updates, or if it does **and** you have created a separate PR in [dvc.org](https://github.com/iterative/dvc.org) with such updates (or at least opened an issue about it in that repo). Please link below to your PR (or issue) in the [dvc.org](https://github.com/iterative/dvc.org) repo.

* [ ] ❌ Have you checked DeepSource, CodeClimate, and other sanity checks below? We consider their findings recommendatory and don't expect everything to be addressed. Please review them carefully and fix those that actually improve code or fix bugs.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Suor commented on dvc/data_cloud.py:
```diff
@@ -39,7 +39,9 @@ def get_remote(self, remote=None, command="<command>"):
             return self._init_remote(remote)
 
-        raise NoRemoteError(command)
+        has_any_remote = any(key.startswith("remote ") for key in self._config)
```
This feels as a part of the logic inside exception itself. So everything should be either inside or outside.

Also, `key.startswith("remote ")` is a bit fishy, we check for regex in other places.

fabiosantoscode replied:
Got it, changed that to use `list_options` as I've found elsewhere. Remote URLs are required so finding any means that there's a valid remote that could be used.

pared commented on dvc/config.py:
```diff
@@ -24,13 +24,29 @@ def __init__(self, msg):
-            "or use:\n"
-            "    dvc {} -r <name>\n".format(command)
-        )
+    def __init__(self, command, *, config=None):
```
what is '*' for?

fabiosantoscode replied:
It's to make sure `config` can only be used as a keyword argument.

fabiosantoscode replied:
The reasoning for it is to prevent the exception from being initialized with some context that might not be a Config.

pared replied:
So it is to make sure that we will never use it as a keyword argument? Isn't it kind of protecting the code from the developer?

fabiosantoscode replied:
Exactly.

pared replied:
Sorry, yeah, I meant that its to make sure that we always use it as a keyword argument.

pared replied:
@fabiosantoscode but I still do not understand why would we need that here, what makes `config` so special that it needs to be always used as keyword argument?

fabiosantoscode replied:
Nothing makes it special really. I just thought somebody might make a mistake and pass something else as an argument here. But I'll change it back.

pared replied:
@fabiosantoscode this is a good point, and we actually have seen few times when wrong arguments are passed due to ordering, but I guess that there is no point in introducing that in this single place if we do differently everywhere else. We just need to be careful when using `kwargs` as `args`.

Suor replied:
I am +1 on using keyword only args in situations like this.

pared replied:
@Suor  So maybe we should enforce this rule throughout the code?

efiop replied:
@pared @Suor I think we have a ticket for doing that in api.py already. Just need to be consistent, we weren't doing that earlier because py2 didn't support it, IIRC.

Suor commented:
Approved it.

A general consideration regarding messages like this - we've discussed adding more or less uniform HINTs in such situations, which will take into account context better, e.g. remote(s) name, having several remotes, command/options/api call. How this should be implemented and how far should we go is still TBD in #2559.

Suor commented:
Thx, @fabiosantoscode.

efiop commented on dvc/config.py:
```diff
@@ -24,13 +24,29 @@ def __init__(self, msg):
+                "    dvc remote add <remote name> <remote url>\n"
+                "To create a default remote use the -d flag:\n"
+                "    dvc remote add -d <remote name> <remote url>"
+            )
```
Having such a complex logic in the exception is self is rather odd. Why not process config in the code itself and then raise NoRemoteError with the pre-crafted message?

fabiosantoscode replied:
Works for me. The message was being generated in `__init__` so I kept it there.

efiop commented on dvc/config.py:
```diff
@@ -24,13 +24,29 @@ def __init__(self, msg):
+        else:
+            msg = (
+                "no remote specified. Create a remote with\n"
+                "    dvc remote add <remote name> <remote url>\n"
```
This part is no relevan without `-d`, because if someone would've provided the remove, we would go `if remote` branch of execution (see below where you raise). So the hint here should assume `-d`.

fabiosantoscode replied:
That's right!

fabiosantoscode commented:
@efiop I've tried to address your comments with a new commit. I've removed the `dvc remote add` without the `-d` and moved the message formatting out of the exception class.
